apps/src/aiComponentLibrary/chatMessage/ChatMessage.tsx

```diff
@@ -1,75 +1,108 @@
   chatMessageText,
   role,
   status,
+  isTeacherView,
 }) => {
+  const [profaneMessageViewToggle, setProfaneMessageViewToggle] = useState(
+    ProfaneMessageViewToggle.VIEW
+  );
+
   const hasDangerStyle =
     status === Status.PROFANITY_VIOLATION ||
     (role === Role.ASSISTANT && status === Status.ERROR);
 
   const hasWarningStyle = status === Status.PII_VIOLATION;
 
+  const getDisplayText: string = useMemo(() => {
+    switch (status) {
+      case Status.OK:
+      case Status.UNKNOWN:
+        return chatMessageText;
+      case Status.PROFANITY_VIOLATION:
+        return role === Role.USER &&
+          profaneMessageViewToggle === ProfaneMessageViewToggle.HIDE
+          ? chatMessageText
+          : commonI18n.aiChatInappropriateUserMessage();
+      case Status.PII_VIOLATION:
+        return commonI18n.aiChatTooPersonalUserMessage();
+      case Status.ERROR:
+        return role === Role.ASSISTANT
+          ? commonI18n.aiChatResponseError()
+          : chatMessageText;
+      default:
+        return '';
+    }
+  }, [chatMessageText, role, status, profaneMessageViewToggle]);
+
   return (
-    <div className={moduleStyles[`container-${role}`]}>
-      {role === Role.ASSISTANT && (
-        <div className={moduleStyles.botIconContainer}>
-          <img
-            src={aiBotIcon}
-            alt={commonI18n.aiChatBotIconAlt()}
-            className={moduleStyles.botIcon}
-          />
-        </div>
-      )}
-      <div
-        className={classNames(
-          moduleStyles[`message-${role}`],
-          hasDangerStyle && moduleStyles.danger,
-          hasWarningStyle && moduleStyles.warning
+    <>
+      <div className={moduleStyles[`container-${role}`]}>
+        {role === Role.ASSISTANT && (
+          <div className={moduleStyles.botIconContainer}>
+            <img
+              src={aiBotIcon}
+              alt={commonI18n.aiChatBotIconAlt()}
+              className={moduleStyles.botIcon}
+            />
+          </div>
         )}
-      >
-        <SafeMarkdown
-          markdown={getDisplayText(chatMessageText, status, role)}
-        />
+        <div
+          className={classNames(
+            moduleStyles[`message-${role}`],
+            hasDangerStyle && moduleStyles.danger,
+            hasWarningStyle && moduleStyles.warning
+          )}
+        >
+          <SafeMarkdown markdown={getDisplayText} />
+        </div>
       </div>
-    </div>
+      {isTeacherView &&
+        role === Role.USER &&
+        status === Status.PROFANITY_VIOLATION && (
+          <div className={moduleStyles[`container-user`]}>
+            <Button
+              onClick={() => {
+                if (
+                  profaneMessageViewToggle === ProfaneMessageViewToggle.VIEW
+                ) {
+                  setProfaneMessageViewToggle(ProfaneMessageViewToggle.HIDE);
+                } else {
+                  setProfaneMessageViewToggle(ProfaneMessageViewToggle.VIEW);
+                }
```

Comment:
I must've missed commenting on this...another suggestion as opposed to a necessity, but this is the sort of thing I'd tend to put in a ternary:

```
setProfaneMessageViewToggle(
  profaneMessageViewToggle  === ProfaneMessageViewToggle.VIEW
    ? ProfaneMessageViewToggle.HIDE
    : ProfaneMessageViewToggle.HIDE
);
```

Main benefit is to not have the function call in two places.


And, further waxing philosophical as opposed to making a serious suggestion here, state machines can work great for this sort of thing.

```
// defined..somewhere
const nextMessageState = {
  ProfaneMessageViewToggle.HIDE : ProfaneMessageViewToggle.VIEW,
  ProfaneMessageViewToggle.VIEW : ProfaneMessageViewToggle.HIDE,
}

// and then in the onClick
  onClick = { () => setProfaneMessageViewToggle( nextMessageState[profaneMessageViewToggle] ) }
```

Since we've only got two states I think sticking with the ternary is smart, this strategy is really most useful for when there are more states. I just figured I'd jot down the concept somewhere. 😄

Reply:
Oh cool! Great for future reference.

Reply:
Will update to ternary.